toolchain/check/import_cpp.cpp

```diff
@@ -44,6 +44,82 @@ static auto GenerateCppIncludesHeaderCode(
   return code;
 }
 
+namespace {
+
+// Used to convert Clang diagnostics to Carbon diagnostics.
+class CarbonClangDiagnosticConsumer : public clang::DiagnosticConsumer {
+ public:
+  // Creates an instance with the location that triggers calling Clang.
+  CarbonClangDiagnosticConsumer(Context& context, SemIRLoc loc)
+      : context_(context), loc_(loc) {}
+
+  // Generates a Carbon warning for each Clang warning and a Carbon error for
+  // each Clang error or fatal.
+  auto HandleDiagnostic(clang::DiagnosticsEngine::Level diag_level,
+                        const clang::Diagnostic& info) -> void override {
+    DiagnosticConsumer::HandleDiagnostic(diag_level, info);
+
+    llvm::SmallString<256> message;
+    info.FormatDiagnostic(message);
+
+    RawStringOstream diagnostics_stream;
+    clang::TextDiagnostic text_diagnostic(
+        diagnostics_stream,
+        // TODO: Consider allowing setting `LangOptions` or use
+        // `ASTContext::getLangOptions()`.
+        clang::LangOptions(),
+        // TODO: Consider allowing setting `DiagnosticOptions` or use
+        // `ASTUnit::getDiagnostics().::getLangOptions().getDiagnosticOptions()`.
+        new clang::DiagnosticOptions());
+    text_diagnostic.emitDiagnostic(
+        clang::FullSourceLoc(info.getLocation(), info.getSourceManager()),
+        diag_level, message, info.getRanges(), info.getFixItHints());
+
+    std::string diagnostics_str = diagnostics_stream.TakeStr();
+
+    switch (diag_level) {
+      case clang::DiagnosticsEngine::Ignored:
+      case clang::DiagnosticsEngine::Note:
+      case clang::DiagnosticsEngine::Remark: {
+        context_.TODO(
+            loc_, llvm::formatv(
+                      "Unsupported: C++ diagnostic level for diagnostic\n{0}",
+                      diagnostics_str));
+        return;
+      }
+      case clang::DiagnosticsEngine::Warning:
+      case clang::DiagnosticsEngine::Error:
+      case clang::DiagnosticsEngine::Fatal: {
+        CARBON_DIAGNOSTIC(CppInteropParseWarning, Warning, "C++:\n{0}",
+                          std::string);
+        CARBON_DIAGNOSTIC(CppInteropParseError, Error, "C++:\n{0}",
+                          std::string);
+        CARBON_DIAGNOSTIC(InCppImport, Note, "in `Cpp` import");
+
+        // TODO: Use a more specific location.
+        context_.emitter()
+            .Build(SemIR::LocId::None,
+                   diag_level == clang::DiagnosticsEngine::Warning
+                       ? CppInteropParseWarning
+                       : CppInteropParseError,
+                   diagnostics_str)
+            .Note(loc_, InCppImport)
+            .Emit();
+        return;
+      }
+    }
+  }
+
+ private:
+  // The type-checking context in which we're running Clang.
+  Context& context_;
```

Comment:
```suggestion
  Context* context_;
```

Style: "When storing an object's address as a non-owned member, prefer storing a pointer."

Reply:
Thanks! I actually prefer pointers in these cases but saw that it's very popular to keep references in Carbon codebase.
Done.